examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -28,144 +25,378 @@
-        self.picks = picks
+        self.sort_by_spectral_ratio = sort_by_spectral_ratio
+        if n_fft is None:
+            self.n_fft = int(sampling_freq)
+        else:
+            self.n_fft = int(n_fft)
+        self.picks_ = picks
+        self.return_filtered = return_filtered
         self.estimator = estimator
         self.n_components = n_components
-
-    def fit(self, inst):
-        """Fit"""
-        if self.picks == None:
-            self.picks_ = mne.pick_types(inst.info, meg=True, eeg=False, ref_meg=False)
-        else:
-            self.picks_ = self.picks
-        if isinstance(inst, mne.io.base.BaseRaw):
+        self.rank = rank
+        self.sampling_freq = sampling_freq
+        self.cov_method_params = cov_method_params
+
+    def _check_X(self, inst):
+        """Check input data."""
+        if not isinstance(inst, np.ndarray):
+            raise ValueError("X should be of type ndarray (got %s)."
+                             % type(inst))
+        if inst.ndim < 3:
+            raise ValueError('X must have at least 3 dimensions.')
+
+    def filter_data(self, inst):
+        if isinstance(inst, BaseRaw):
             inst_signal = inst.copy()
             inst_signal.filter(picks=self.picks_, **self.filt_params_signal)
-
+            # noise
             inst_noise = inst.copy()
             inst_noise.filter(picks=self.picks_, **self.filt_params_noise)
-            if self.subtract_signal:
-                inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
-            cov_signal = mne.compute_raw_covariance(
-                inst_signal, picks=self.picks_, method=self.estimator)
-            cov_noise = mne.compute_raw_covariance(inst_noise, picks=self.picks_,
-                method=self.estimator)
+            # subtract signal:
+            inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
+        else:
+            if isinstance(inst, BaseEpochs) or isinstance(inst, np.ndarray):
+                # data X is epoched
+                # part of the following code is copied from mne csp
+                n_epochs, n_channels, n_samples = inst.shape
+                self.max_components = n_channels
+                # reshape for filtering
+                X_aux = np.reshape(inst, [n_epochs, n_channels*n_samples])
+                inst_signal = filter_data(
+                    X_aux, self.sampling_freq, **self.filt_params_signal)
+                # rephase for filtering
+                X_aux = np.reshape(inst, [n_epochs, n_channels*n_samples])
+                inst_noise = filter_data(
+                    X_aux, self.sampling_freq, **self.filt_params_noise)
+                # subtract signal:
+                inst_noise -= inst_signal
+                # Estimate single trial covariance
+                # reshape to original shape
+                inst_signal = np.reshape(
+                    inst_signal, [n_epochs, n_channels, n_samples])
```

Comment:
Are line breaks necessary here in terms of 80 characters? alsop below in 214

Reply:
yes! It has 82 characters !